Objects/iterobject.c

```diff
@@ -288,3 +288,323 @@ PyTypeObject PyCallIter_Type = {
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *wrapped_iter; /*
+        The iterator returned by the callable, unwrapped via __await__.
+
+        If NULL this means that the iterator is already exhausted and when
+        iterated it should raise StopAsyncIteration.
+    */
+    callasynciterobject *it; /* The iterator object, in order to clear it when done. */
+} callasynciterawaitableobject;
+
+
+PyObject *
+PyCallAsyncIter_New(PyObject *callable, PyObject *sentinel)
+{
+    callasynciterobject *it;
+    it = PyObject_GC_New(callasynciterobject, &PyCallAsyncIter_Type);
+    if (it == NULL)
+        return NULL;
+    Py_INCREF(callable);
+    it->it_callable = callable;
+    Py_INCREF(sentinel);
+    it->it_sentinel = sentinel;
+    _PyObject_GC_TRACK(it);
+    return (PyObject *)it;
+}
+
+static void
+callasynciter_dealloc(callasynciterobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->it_callable);
+    Py_XDECREF(obj->it_sentinel);
+    PyObject_GC_Del(obj);
+}
+
+static void
+callasynciterawaitable_dealloc(callasynciterawaitableobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->wrapped_iter);
+    Py_XDECREF(obj->it);
+    PyObject_GC_Del(obj);
+}
+
+static int
+callasynciter_traverse(callasynciterobject *it, visitproc visit, void *arg)
+{
+    Py_VISIT(it->it_callable);
+    Py_VISIT(it->it_sentinel);
+    return 0;
+}
+
+static int
+callasyncawaitable_traverse(callasynciterawaitableobject *obj, visitproc visit, void *arg)
+{
+    Py_VISIT(obj->wrapped_iter);
+    Py_VISIT(obj->it);
+    return 0;
+}
+
+static PyObject *
+callasynciter_anext(callasynciterobject *it)
+{
+    PyObject *obj = NULL, *wrapped_iter = NULL;
+    PyTypeObject *t = NULL;
+    callasynciterawaitableobject *awaitable;
+
+    if (it->it_callable == NULL) {
+        /* Can we raise this at this point, or do we need to return an awaitable
+         * that raises it? */
+        PyErr_SetNone(PyExc_StopAsyncIteration);
+        return NULL;
+    }
+
+    obj = _PyObject_CallNoArg(it->it_callable);
+    if (obj == NULL) {
+        return NULL;
+    }
+
+    t = Py_TYPE(obj);
+    if (t->tp_as_async == NULL || t->tp_as_async->am_await == NULL) {
+        return PyErr_Format(PyExc_TypeError, "'%.200s' object is not awaitable", obj);
+    }
+    
+    wrapped_iter = (*t->tp_as_async->am_await)(obj);
+
+    awaitable = PyObject_GC_New(
+        callasynciterawaitableobject,
+        &PyAsyncCallAwaitable_Type);
+    if (awaitable == NULL) {
+        return NULL;
+    }
+
+    Py_INCREF(it);
+    awaitable->it = it;
+    awaitable->wrapped_iter = wrapped_iter;
+
+    _PyObject_GC_TRACK(awaitable);
+    return (PyObject *) awaitable;
+}
+
+static PyObject *
+asynccallawaitable_iternext(callasynciterawaitableobject *obj)
+{
+    PyObject *result;
+    PyObject *stop_value = NULL;
+    int ok;
+
+    if (obj->it->it_sentinel == NULL) {
+        return PyErr_Format(PyExc_TypeError, "'%.200s' object is already exhausted", obj->it);
```

Comment:
Shouldn't this be `AsyncStopIteration`? That seems to be how `next()` behaves on callables

Reply:
Makes sense to me - @lordmauve we don't add this error message if we raise `StopAsyncIteration` right? Formatting with `obj->it` doesn't seem to work anyways, think we would need `Py_TYPE(obj->it)->tp_name`

Reply:
Done in 06019ea05ebcd477f1545f2fccc8b089204ade42. I think this can be resolved.